File: src/main/kotlin/com/github/bin/command/RegexCommand.java
package com.github.bin.command;

import com.github.bin.service.RoomConfig;
import lombok.RequiredArgsConstructor;
import lombok.val;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author bin
 * @since 2023/08/22
 */
@RequiredArgsConstructor
public abstract class RegexCommand implements Command {
    private final Pattern regex;

    @Override
    public final boolean invoke(RoomConfig roomConfig, String id, String msg) {
        val result = regex.matcher(msg);
        if (result.find()) {
            return handler(roomConfig, id, result);
        }
        return false;
    }

    abstract boolean handler(RoomConfig roomConfig, String id, Matcher msg);
}

File: src/main/kotlin/com/github/bin/command/RegexCommand.kt
package com.github.bin.command

import com.github.bin.service.RoomConfig

/**
 * @author bin
 * @version 1.0.0
 * @since 2023/7/9
 */
abstract class RegexCommand(
        private val regex: Regex
) : Command {
    override fun invoke(roomConfig: RoomConfig, id: String, msg: String): Boolean {
        val result = test(msg) ?: return false
        return handler(roomConfig, id, result)
    }

    open fun test(msg: String): MatchResult? {
        return regex.find(msg)
    }

    abstract fun handler(roomConfig: RoomConfig, id: String, msg: MatchResult): Boolean
}
